adr/ADR-21.md

```diff
@@ -0,0 +1,119 @@
+
+The intention of the ADR is to document the storage of these contexts so that clients can, optionally, support using them.
+
+## Version History
+
+|Date|Revision|
+|----|--------|
+|2020-08-12|Initial basic design|
+|2020-05-07|JetStream Domains|
+|2021-12-13|Custom Inbox Prefix|
+
+This reflects a current implementation in use widely via the CLI as such it's a stable release.  Only non breaking additions will be considered.
+
+## Design
+
+Today the design is entirely file based for maximum portability, later we can consider other options like S3 buckets, KV stores etc.
+
+### Configuration Paths
+
+There is generally no standard for what goes on in a users home directory on a Unix system. Recently the Free Desktop team have been working on [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) that specifies in detail where configuration, data, binaries and more are to be stored in a way that's compatible with Linux desktops like KDE and Gnome but also with systems such as systemd.
+
+We therefore based the design on this specification as a widely supported standard.
+
+|File|Description|
+|----|-----------|
+|`~/.config`|The default location for user configuration, configurable using `XDG_CONFIG_HOME`|
+|`~/.config/nats`|Where all NATS related user configuration should go|
+|`~/.config/nats/context.txt`|The current selected (default) context, would contain just `ngs`|
+|`~/.config/nats/context/ngs.json`|The configuration for the `ngs` context|
+
+While this is Linux centered it does work on Windows, we might want to consider a more typical path to replace `~/.config` there and keep the rest as above.
+
+### Context content
+
+The `~/.config/nats/context/ngs.json` file has the following JSON fields:
+
+|Key|Default|Description|
+|---|-------|-----------|
+|`description`| |A human friendly description for the specific context|
+|`url`|`nats://localhost:4222`|Comma seperated list of server urls|
+|`token`| |Authentication token|
+|`user`| |The username to connect with, requires a password|
+|`password`| |Password to connect with|
+|`creds`| |Path to a NATS Credentials file|
+|`nkey`| |Path to a NATS Nkey file|
+|`cert`| |Path to the x509 public certificate|
+|`key`| |Path to the x509 private key|
+|`ca`| |Path to the x509 Certificate Authority|
+|`nsc`| |A `nsc` resolve url for loading credentials and server urls|
+|`jetstream_domain`| |The JetStream Domain to use|
+|`jetstream_api_prefix`| |The JetStream API Prefix to use|
+|`jetstream_event_prefix`| |The JetStream Event Prefix|
+|`inbox_prefix`| |A prefix to use when generating inboxes|
+
+All fields are optional, none are marked as `omitempty`, users wishing to edit these with an editor should known all valid key names.
+
+Above settings map quite obviously to client features with the exception of `nsc`, the `nsc` key takes a URL like value, examples are:
```

Comment:
Should that nsc be invoked with store and nkey directory specified by the context as well?

Reply:
Quite happy to let those who choose to customise everything carry that  complexity burden themselves by just pointing at the credential file themselves. 

This infinite customizability we are adding at every level is a plague. 

Why would nsc need a context? @aricart and I discussed that long ago and decided it’s not compatible models

Reply:
Also, afaik, nsc reads environment variables and a config file, not options you pass? `NSC_HOME` and `NKEYS_PATH` etc, I am not going to be supporting passin these around. If someone sets them they will continue to work for nsc when invoked by the context.